### ISSUE TITLE
Feature request: Deep pick

### ISSUE THREAD

**Issue by zatziky** [ramda/ramda#3012]
Hello guys,

do you plan to add a `pickDeep()` function? Or do you have a suggestion on how to achieve this functionality?

**Deep pick** is useful when you need only an exact part object. For example:

```
const fullObject = {
   a: 1,
   b: {
     c: 2,
     d: 3
   }
}

const partialObject = pickDeep(['b', 'c'], fullObject)
// OR multiarray version to pick multiple parts: `pickDeep([['b', 'c']], fullObject)`

expect(partialObject).to.eql({
   b: {
     c: 2
   }
})
```

**Comment by CrossEye** [comment]
While I don't see this being added to Ramda in the near future, feel free to add it to the [Cookbook][cb].  It's an interesting feature, but this is the first time I've seen any such request.  It's not clear to me that there is a widespread need fot it.

Here are possible implementations for the two APIs you suggested:

```js
const pickDeep = curry (
  (pth, obj) => assocPath (pth, path (pth, obj), {})
)
```

```js
const pickDeep = curry (
  (pths, obj) => reduce ((acc, pth) => assocPath (pth, path (pth, obj), acc), {}, pths)
)
```

You might want to find better parameter names than `pth` or `pths`.  :smile:  Also, this is somewhat naive in that it always starts with a plain object, but Ramda would offer you the ability to start with an array if the first term in the path is a non-negative integer rather than a string.  You would just need to replace the initial value of `{}` with `[]` for those cases.  But that's left as an exercise for the reader.

Feel free to try to convince us that this is important enough to include.  My first thought is no, but that could change with a convincing argument.

  [cb]: https://github.com/ramda/ramda/wiki/Cookbook

**Comment by zatziky** [comment]
Hey Scott, @CrossEye 

thank you for the proposed solution. An argument you need. So what about these people who already needed or need such a feature? ;-)
- https://www.npmjs.com/package/pick-deep
- https://glebbahmutov.com/blog/deep-picking/

In Amio, we will be using it extensively now with Hapi Joi validation framework.

**Comment by CrossEye** [comment]
As I said, the function is interesting.  I'm not surprised people have already written solutions.  The question is whether it belongs in Ramda proper.  I'm afraid Ramda is already seeming a bit bloated.  So a function really needs to be either (a) a very widespread need or (b) difficult to  implement well on one's own.  I'm not seeing either.

One other suggestion: as well as the cookbook there are great projects such as [ramda-adjunct][ra] and [ramda extension][re] which offer functions which, while related, have never managed to make it into Ramda itself.  Their authors might well consider adding this.


  [ra]: https://github.com/char0n/ramda-adjunct
  [re]: https://github.com/tommmyy/ramda-extension

**Comment by zatziky** [comment]
@CrossEye Scott,

> I'm afraid Ramda is already seeming a bit bloated

What do you mean by bloated? 
Like in size? It should not be a problem as you can export exactly whatever method you need via `ramda/src/*`
Or by the sheer number of features? I believe that each feature should have the modifiers just like the [drop()](https://ramdajs.com/docs/#drop):
- drop
- dropLast
- dropLastWhile
- dropRepeats
- dropRepeatsWith
- dropWhile

What I am saying, do not only add `pickDeep()` but consider being more open in adding other alternatives too. Just a humble opinion as a long term ramda user. 😉

Thanks for the 2 libs. I didn't know them.

**Comment by CrossEye** [comment]
By bloated, I simply mean that the library was never meant to be a one-stop-shopping center for all utility functions one might need.  Instead it was supposed to be a carefully cultivated list of functions that made it easy for users to build what they need.  When we stray beyond that, we add to the bloat.

I'm still not convinced.  Anyone else want to chime in?  @ramda/core?

**Comment by Bradcomp** [comment]
I have heard more complaints that there are too many functions in Ramda than too few. Lots of "Oh, I didn't know I could do that!" type stuff. When the api gets bigger it's more for people to know. This isn't a big deal but on the other hand we have the "what function should I use" section of the wiki so I know it's a problem people run into.

I argued against exploding our `merge*` api, I am on the same side of the fence for this one. There are lots of different ways to get similar functionality from the library, plus the cookbook and supplemental libraries.

**Comment by zatziky** [comment]
I wonder how bad is this "Oh, I didn't know I could do that!"? 

An API can grow bigger but a reasonable categorization should help. Categorization can even be done by proper naming (which Ramda follows). Being consistent in function naming and the attached modifier, not much of the functionality will be hidden from users.

E.g. you need a merge function, so you look for `merge*`. Then users will get familiar with the "modifiers" and because of consistency they will know how to apply them with other functions. The "modifiers" like `All`, `DeepLeft`, `DeepRight`.  From `merge` + `All`, you can then easily get to `insert` + `All`. 

Basically, what I'm saying **the modifiers are not making the API bigger****, it is the different functions**.

**Comment by CrossEye** [comment]
@zatziky:

> An API can grow bigger but a reasonable categorization should help. 

I wonder if you have suggestions for how Ramda might do this better.  I don't think we do a very good job right now.

**Comment by zatziky** [comment]
@CrossEye Scott, my personal experience with searching for a function in Ramda is really good. The search bar plus the naming conventions are already working great!

![image](https://user-images.githubusercontent.com/2794365/84113896-b1b58380-aa2b-11ea-8aff-1f4823dae574.png)

In this search, I know that I need a `take`. I start searching for **take** and voila ramda-docs show me all possible modifiers. Since you are using the modifier **last** consistently, I can use `takeLast` without even having to read the documentation since recently I used `dropLast`. The search bar jus helped me discover it.

You are (ramda.js team) maybe just scarred from some negative comments. I'd say, you already have a very easy to "read and search" documentation.
